config/provider.go

```diff
@@ -91,3 +95,22 @@ func formatResourceNames(resources [][]string) []string {
 
 	return formatted
 }
+
+func terraformProvider() *schema.Provider {
+	p := upcloud.Provider()
+	p.ConfigureContextFunc = func(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
+		return upcloud.ProviderConfigure(ctx, d, defaultUserAgent())
+	}
+
+	return p
+}
+
+func terraformPluginFrameworkProvider() provider.Provider {
+	p := upcloud.NewWithUserAgent(defaultUserAgent())
+
+	return p
+}
+
+func defaultUserAgent() string {
+	return fmt.Sprintf("provider-upcloud/%s", version)
```

Comment:
Should the user-agent include `crossplane`?

```suggestion
	return fmt.Sprintf("crossplane-provider-upcloud/%s", version)
```

Reply:
The prefix addition is included in https://github.com/UpCloudLtd/provider-upcloud/pull/13 , but better to have it here already. :+1: